plugin-viewport/src/main/kotlin/com/mapbox/maps/plugin/viewport/experimental/state/MapboxViewportStateBase.kt

```diff
@@ -0,0 +1,156 @@
+   */
+  override fun transitionToState(
+    stateTransitionOptions: ViewportTransitionOptions,
+    frameTransitionOptions: ViewportTransitionOptions,
+    transitionEndListener: TransitionEndListener?
+  ) {
+    dataSource.registerUpdateObserver(dataSourceUpdateObserver)
+    startAnimation(
+      stateTransition.transitionToState(dataSource.getViewportData().camera, stateTransitionOptions)
+        .apply {
+          addListener(
+            createTransitionListener(frameTransitionOptions)
+          )
+        },
+      instant = false,
+      transitionEndListener
+    )
+  }
+
+  /**
+   * Clean up current state, cancel all the ongoing viewport animations.
+   */
+  override fun cleanUp() {
+    cancelAnimation()
+    dataSource.unregisterUpdateObserver(dataSourceUpdateObserver)
+
+  }
+
+  private fun cancelAnimation() {
+    runningAnimation?.apply {
+      cancel()
+      childAnimations.forEach {
+        cameraPlugin.unregisterAnimators(it as ValueAnimator)
+      }
+    }
+    runningAnimation = null
+  }
+
+  private fun startAnimation(
+    animatorSet: AnimatorSet,
+    instant: Boolean,
+    transitionEndListener: TransitionEndListener? = null,
+  ) {
+    cancelAnimation()
+    transitionEndListener?.let {
+      transitionEndListeners.add(it)
+    }
+    animatorSet.childAnimations.forEach {
+      cameraPlugin.registerAnimators(it as ValueAnimator)
+    }
+    if (instant) {
+      animatorSet.duration = 0
+    }
+    animatorSet.start()
+    runningAnimation = animatorSet
+  }
+
+  private fun finishAnimation(animatorSet: AnimatorSet?) {
+    animatorSet?.childAnimations?.forEach {
+      cameraPlugin.unregisterAnimators(it as ValueAnimator)
+    }
+    if (runningAnimation == animatorSet) {
+      runningAnimation = null
+    }
+  }
+
+  private fun updateFrame(viewportData: ViewportData, instant: Boolean = false) {
+    startAnimation(
+      stateTransition.updateFrame(
+        viewportData.camera,
+        frameTransitionOptions
+      ).apply { doOnEnd { finishAnimation(this) } },
+      instant
+    )
+  }
+
+
+  private fun createTransitionListener(
+    frameTransitionOptions: ViewportTransitionOptions,
+  ) = object : Animator.AnimatorListener {
+
+    private var isCanceled = false
+
+    override fun onAnimationStart(animation: Animator?) {
+      this@MapboxViewportStateBase.frameTransitionOptions = ViewportPlugin.DEFAULT_FRAME_TRANSITION_OPT
+      this@MapboxViewportStateBase.isTransitioning = true
+    }
+
+    override fun onAnimationEnd(animation: Animator?) {
+      this@MapboxViewportStateBase.isTransitioning = false
+      if (!isCanceled) {
+        this@MapboxViewportStateBase.frameTransitionOptions = frameTransitionOptions
+      }
+
+      finishAnimation(animation as AnimatorSet)
+      transitionEndListeners.forEach { it.onTransitionEnd(isCanceled) }
+      transitionEndListeners.clear()
+      updateFrame(dataSource.getViewportData())
```

Comment:
This should only happen if not canceled.

Reply:
This part of the codebase has been refactored, please review again.